<a href="https://colab.research.google.com/github/philipsarkis/461pFinalProject/blob/main/dsp_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
%%bash
# Install and update essential libraries
apt -q update
apt install -y -q tree


Get:1 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]
Get:2 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease [3,626 B]
Ign:3 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease
Ign:4 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease
Get:5 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release [697 B]
Hit:6 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release
Get:7 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release.gpg [836 B]
Get:8 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic InRelease [15.9 kB]
Hit:9 http://archive.ubuntu.com/ubuntu bionic InRelease
Get:10 http://archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]
Hit:11 http://ppa.launchpad.net/cran/libgit2/ubuntu bionic InRelease
Get:12 http://security.ubuntu.com/ubuntu bionic-securi

In [ ]:
# Download a file from Google Drive using FILE_ID

# Install the PyDrive wrapper & import libraries.
# This only needs to be done once per notebook.
!pip install -U -q PyDrive
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials

# Authenticate and create the PyDrive client.
# This only needs to be done once per notebook.
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)
# Download a file based on its file ID.
#
# A file ID looks like: 1rhnzJ8CD9Amdz5OMi8s4kqT5b00jY5k4
# Visit https://help.meiro.io/en/articles/2245027-where-can-i-find-the-file-id-on-google-drive
# for more details on getting the FILE ID.

def download_file(file_id, file_name):
    '''
    file_id: Make sure that you can access this file from your account or else, this will not run
    file_name: Name of the file by which you want to save
    '''
    downloaded = drive.CreateFile({'id': file_id})
    downloaded.GetContentFile(file_name)
    print('Downloaded file with ID {} and name {}'.format(file_id, file_name))
  
def upload_file(file_name):
    # If the file is large, upload to Google Drive to download easily.
    # Upload a file from the current VM.

    uploaded = drive.CreateFile({'title': file_name})
    uploaded.SetContentFile(file_name)
    uploaded.Upload()
    print('Uploaded file with ID {}'.format(uploaded.get('id')))

In [ ]:
download_file('1hDOg7vrsVb6gvpH6qM84tG2Fwz1ndIpE', 'archive.zip')

Downloaded file with ID 1hDOg7vrsVb6gvpH6qM84tG2Fwz1ndIpE and name archive.zip


In [ ]:
!unzip 'drinve/MyDrive/archive.zip'

Archive:  drive/MyDrive/archive.zip
replace IMDB_movie_details.json? [y]es, [n]o, [A]ll, [N]one, [r]ename: 

In [ ]:
!unzip 'archive.zip'

Archive:  archive.zip
  inflating: IMDB_movie_details.json  
  inflating: IMDB_reviews.json       


In [ ]:
import pandas as pd

In [ ]:
reviews = pd.read_json('IMDB_reviews.json',lines=True)
reviews.head()

review_date  ...                                 review_summary
0  10 February 2006  ...  A classic piece of unforgettable film-making.
1  6 September 2000  ...     Simply amazing. The best film of the 90's.
2     3 August 2001  ...               The best story ever told on film
3  1 September 2002  ...                     Busy dying or busy living?
4       20 May 2004  ...         Great story, wondrously told and acted

[5 rows x 7 columns]

In [ ]:
movies = pd.read_json('IMDB_movie_details.json',lines=True)
movies.head()

movie_id  ...                                      plot_synopsis
0  tt0105112  ...  Jack Ryan (Ford) is on a "working vacation" in...
1  tt1204975  ...  Four boys around the age of 10 are friends in ...
2  tt0243655  ...                                                   
3  tt0040897  ...  Fred Dobbs (Humphrey Bogart) and Bob Curtin (T...
4  tt0126886  ...  Jim McAllister (Matthew Broderick) is a much-a...

[5 rows x 7 columns]

In [ ]:
test = pd.DataFrame()
test = test.append(movies.loc[movies['movie_id'] == 'tt0105112'])
test.head()

movie_id  ...                                      plot_synopsis
0  tt0105112  ...  Jack Ryan (Ford) is on a "working vacation" in...

[1 rows x 7 columns]

In [ ]:
movie_ids = reviews['movie_id']
to_append = pd.DataFrame()
j = 0
for i in movie_ids:
  to_append = to_append.append(movies.loc[movies['movie_id'] == i])
  if j%5000 == 0:
    print(j)
  j+= 1
to_append.head()

0
5000
10000
15000
20000
25000
30000
35000
40000
45000
50000
55000
60000
65000
70000
75000
80000
85000
90000
95000
100000
105000
110000
115000
120000
125000
130000
135000
140000
145000
150000
155000
160000
165000
170000
175000
180000
185000
190000
195000
200000
205000
210000
215000
220000
225000
230000
235000
240000
245000
250000
255000
260000
265000
270000
275000
280000
285000
290000
295000
300000
305000
310000
315000
320000
325000
330000
335000
340000
345000
350000
355000
360000
365000
370000
375000
380000
385000
390000
395000
400000
405000
410000
415000
420000
425000
430000
435000
440000
445000
450000
455000
460000
465000
470000
475000
480000
485000
490000
495000
500000
505000
510000
515000
520000
525000
530000
535000
540000
545000
550000
555000
560000
565000
570000


movie_id  ...                                      plot_synopsis
1263  tt0111161  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
1263  tt0111161  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
1263  tt0111161  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
1263  tt0111161  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
1263  tt0111161  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...

[5 rows x 7 columns]

In [ ]:
bbbbbbbbb

In [ ]:
reviews.shape

(573913, 7)

In [ ]:
to_append.head()

movie_id  ...                                      plot_synopsis
1263  tt0111161  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
1263  tt0111161  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
1263  tt0111161  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
1263  tt0111161  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
1263  tt0111161  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...

[5 rows x 7 columns]

In [ ]:
to_append.to_csv("to_append.csv")

In [ ]:
to_append.shape

(573906, 7)

In [ ]:
%%bash

pip install httplib2==0.15.0

In [ ]:
upload_file("append.zip")

RedirectMissingLocation: ignored

In [ ]:
%%bash
zip append.zip to_append.csv

  adding: to_append.csv (deflated 94%)


In [ ]:
file_list = drive.ListFile({'q': "'root' in parents and trashed=false"}).GetList()
for file1 in file_list:       
    print ('title: %s, id: %s' % (file1['title'], file1['id']))
sys.exit()

title: archive.zip, id: 1hDOg7vrsVb6gvpH6qM84tG2Fwz1ndIpE
title: ANS Paper, id: 1GohI4FO1sC7ySz_0lj3VRraCSNPN-i2SwomMk_uu-Sg
title: Kaggle Writeup, id: 1R5K5fLk8khKYTYDfKuMwxBJ8bnGCanl0NcnbVUanE8o
title: DSP Final Proposal, id: 14CMd3Gm0dfohRbo7V1K_re3pDGkczdxJCKHa3HhRKI8
title: train_final.csv, id: 11DRW2sehLt4gHTQsuTtEJf1reTrRkfj9
title: test_final.csv, id: 1dMcohMzDlDzUdX5m8B7Ko4VCaO6xeta-
title: midterm1_Sahil_Vaidya.doc, id: 15ToU9emPJ8nedhbfhJjSVb9_jzcPPYHv
title: Thought Paper 1, id: 1AAHtGgHITsRoyqCivH3-qOelx94gPHAWZ3JPkradqm8
title: Data Notes, id: 1wNBUm6mbVPbUoFoU72zhoeBF2d4ytrs5ECmo8VZhqXY
title: Hendrickson Contract Service Document_Vaidya.xlsx, id: 116d-YhOoQM7RVmuszF9ck09xuKUvsMiW
title: w-9_vaidya.pdf, id: 1s3uiwI4KgbbBwLk191-7-hj7zZozpLyx
title: Untitled Diagram.drawio, id: 1wW6M06d6POUEurgH2VATzWyPt1djco2C
title: Untitled Diagram.drawio, id: 1YpFgaAuWSi2QVgjSgcEdBVDRHTjaRZCW
title: task4.drawio, id: 1-ObdDZP-_fwRGcWX-TgbkaDhll3rIMNa
title: task2.drawio, id: 1SGKon0LaF

NameError: ignored

Day 2

In [ ]:
to_append = pd.read_csv('drive/MyDrive/to_append.csv')
to_append.head()

Unnamed: 0  ...                                      plot_synopsis
0        1263  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
1        1263  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
2        1263  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
3        1263  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
4        1263  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...

[5 rows x 8 columns]

In [ ]:
to_append.shape

(573906, 8)

In [ ]:
reviews.shape

(573906, 7)

In [ ]:
reviews.drop(reviews.tail(7).index,inplace = True)

AttributeError: ignored

In [ ]:
merge = pd.merge(reviews,to_append, left_index=True, right_index=True)
merge.head()

review_date  ...                                      plot_synopsis
0  10 February 2006  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
1  6 September 2000  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
2     3 August 2001  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
3  1 September 2002  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
4       20 May 2004  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...

[5 rows x 15 columns]

In [ ]:
merge.shape

(573906, 15)

In [ ]:
merge.to_csv("dataset.csv")

In [ ]:
%%bash
zip dataset.zip dataset.csv

  adding: dataset.csv (deflated 89%)


In [ ]:
X = merge.drop("is_spoiler", 1)
y = merge['is_spoiler']

In [ ]:
X['movie_id_x'].value_counts()

tt0468569    4845
tt0111161    4361
tt0167260    2729
tt0137523    2480
tt0068646    2137
             ... 
tt0107719      12
tt6294822      11
tt0104014       5
tt0201265       4
tt0114142       2
Name: movie_id_x, Length: 1572, dtype: int64

In [ ]:
len(X['movie_id_x'].unique())

1572

1572 unique movies in dataset, ranging from 5000 reviews to 2 reviews

In [ ]:
X.drop(["user_id","Unnamed: 0"],1,inplace=True)
X.head()

review_date  ...                                      plot_synopsis
0  10 February 2006  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
1  6 September 2000  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
2     3 August 2001  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
3  1 September 2002  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
4       20 May 2004  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...

[5 rows x 12 columns]

In [ ]:
X.drop(["movie_id_y"],1,inplace=True)
X.head()

review_date  ...                                      plot_synopsis
0  10 February 2006  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
1  6 September 2000  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
2     3 August 2001  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
3  1 September 2002  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
4       20 May 2004  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...

[5 rows x 11 columns]

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, roc_auc_score
from sklearn.preprocessing import OneHotEncoder

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [ ]:
X_train_continuous = X_train.drop(["review_date","review_text","review_summary","plot_summary","duration","genre","plot_synopsis", "movie_id_x","release_date"],1)
lr = LogisticRegression()
lr.fit(X_train_continuous,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [ ]:
X_test.shape

(189389, 11)

In [ ]:
X_test_continuous = X_test.drop(["review_date","review_text","review_summary","plot_summary","duration","genre","plot_synopsis", "movie_id_x","release_date"],1)
y_pred = lr.predict(X_test_continuous)
print("accuracy is ", accuracy_score(y_test,y_pred))
print("auc score is ", roc_auc_score(y_test,y_pred))

accuracy is  0.7375771560122288
auc score is  0.5


Can predict with decent accuracy but likely because there are more non-spoilers than spoilers shown by the auc score

In [ ]:
OneHotEncoder().fit_transform(X_train)

ValueError: ignored

In [ ]:
X_train.isna().sum()

review_date           0
movie_id_x            0
review_text           0
rating_x              0
review_summary        0
plot_summary          0
duration              0
genre                 0
rating_y              0
release_date          0
plot_synopsis     23566
dtype: int64

In [ ]:
X_train.fillna('',inplace=True)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4327: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  downcast=downcast,


In [ ]:
X_train.isna().sum()

review_date       0
movie_id_x        0
review_text       0
rating_x          0
review_summary    0
plot_summary      0
duration          0
genre             0
rating_y          0
release_date      0
plot_synopsis     0
dtype: int64

In [ ]:
X_train_enc = pd.DataFrame((OneHotEncoder().fit_transform(X_train)).toarray())

NameError: ignored

In [ ]:
X_train_enc.head()

review_date  ...                                      plot_synopsis
160        4 October 2017  ...  In 1947, Andy Dufresne (Tim Robbins), a banker...
309745     19 August 2014  ...  It's 1995. Jimmy "B-Rabbit" Smith Jr. (Eminem)...
119676  23 September 2010  ...  Set in a surreal world where our whole society...
394727    1 February 2007  ...  The Prestige begins with shots of several doze...
299024       3 March 2001  ...  The film takes place 10 years after the events...

[5 rows x 11 columns]

Day 3

In [ ]:
!unzip 'drive/MyDrive/dataset.zip'

Archive:  drive/MyDrive/dataset.zip
  inflating: dataset.csv             


In [ ]:
df = pd.read_csv("dataset.csv")

In [ ]:
!pip install -q wordcloud

In [ ]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [ ]:
from nltk.corpus import stopwords
stop = stopwords.words('english')

In [ ]:
df['review_text'].apply(lambda x: [item for item in x if item not in stop])

0         [I, n,  ,  , O, c, r,  , e, r, ,,  , S, h, w, ...
1         [T, h, e,  , S, h, w, h, n, k,  , R, e, e, p, ...
2         [I,  , b, e, l, e, v, e,  , h,  , h,  , f, l, ...
3         [*, *, Y, e, ,,  , h, e, r, e,  , r, e,  , S, ...
4         [A,  , h, e,  , h, e, r,  , f,  , h,  , e, x, ...
                                ...                        
573901    [I, ',  , e,  ,  , c, k, n, w, l, e, g, e,  , ...
573902    [I,  , w, e, n,  ,  , h, e,  , v, e,  , l,  , ...
573903    [N,  , b, ,,  , n,  , b,  ,  , l, l, .,  ,  , ...
573904    [I,  , h, n, k,  , h,  , v, e,  , w,  , u, n, ...
573905    [I,  , w,  , G, !,  ,  ,  , k, n,  , w, h,  , ...
Name: review_text, Length: 573906, dtype: object

In [ ]:
df['review_text']

0         In its Oscar year, Shawshank Redemption (writt...
1         The Shawshank Redemption is without a doubt on...
2         I believe that this film is the best story eve...
3         **Yes, there are SPOILERS here**This film has ...
4         At the heart of this extraordinary movie is a ...
                                ...                        
573901    It's time to acknowledge for once that Pulp Fi...
573902    I went to the movies last night, and had the o...
573903    Not bad, not bad at all.  I am a sucker for a ...
573904    I think this movie was under-appreciated by Am...
573905    I saw Go! as a knot with no bows - a very inte...
Name: review_text, Length: 573906, dtype: object

In [ ]:
df['review_without_stopwords'] = df['review_text'].apply(lambda x: ' '.join([word for word in x.split() if word not in (stop)]))

In [ ]:
df['review_without_stopwords']

0         In Oscar year, Shawshank Redemption (written d...
1         The Shawshank Redemption without doubt one bri...
2         I believe film best story ever told film, I'm ...
3         **Yes, SPOILERS here**This film emotional impa...
4         At heart extraordinary movie brilliant indelib...
                                ...                        
573901    It's time acknowledge Pulp Fiction pioneer fil...
573902    I went movies last night, option watching Go W...
573903    Not bad, bad all. I sucker film plays havoc 'p...
573904    I think movie under-appreciated American audie...
573905    I saw Go! knot bows - interesting look teenage...
Name: review_without_stopwords, Length: 573906, dtype: object

In [ ]:
df['description_without_stopwords'] = df['plot_synopsis'].apply(lambda x: ' '.join([word for word in x.split() if word not in (stop)]))

AttributeError: ignored

In [ ]:
df.fillna('',inplace=True)

In [ ]:
df['description_without_stopwords'] = df['plot_synopsis'].apply(lambda x: ' '.join([word for word in str(x).split() if word not in (stop)]))

In [ ]:
df['description_without_stopwords']

0         In 1947, Andy Dufresne (Tim Robbins), banker M...
1         In 1947, Andy Dufresne (Tim Robbins), banker M...
2         In 1947, Andy Dufresne (Tim Robbins), banker M...
3         In 1947, Andy Dufresne (Tim Robbins), banker M...
4         In 1947, Andy Dufresne (Tim Robbins), banker M...
                                ...                        
573901    The film told chronological format, story told...
573902    The film told chronological format, story told...
573903    The film told chronological format, story told...
573904    The film told chronological format, story told...
573905    The film told chronological format, story told...
Name: description_without_stopwords, Length: 573906, dtype: object

In [ ]:
refined_df = df.drop(["user_id","Unnamed: 0","movie_id_x","review_text","plot_synopsis"],1)
refined_df

review_date  ...                      description_without_stopwords
0        10 February 2006  ...  In 1947, Andy Dufresne (Tim Robbins), banker M...
1        6 September 2000  ...  In 1947, Andy Dufresne (Tim Robbins), banker M...
2           3 August 2001  ...  In 1947, Andy Dufresne (Tim Robbins), banker M...
3        1 September 2002  ...  In 1947, Andy Dufresne (Tim Robbins), banker M...
4             20 May 2004  ...  In 1947, Andy Dufresne (Tim Robbins), banker M...
...                   ...  ...                                                ...
573901  11 September 1999  ...  The film told chronological format, story told...
573902   4 September 1999  ...  The film told chronological format, story told...
573903     22 August 1999  ...  The film told chronological format, story told...
573904     18 August 1999  ...  The film told chronological format, story told...
573905     29 August 1999  ...  The film told chronological format, story told...

[573906 rows x 13 columns]

In [ ]:
refined_df = refined_df.drop(["Unnamed: 0.1"],1)
refined_df

review_date  ...                      description_without_stopwords
0        10 February 2006  ...  In 1947, Andy Dufresne (Tim Robbins), banker M...
1        6 September 2000  ...  In 1947, Andy Dufresne (Tim Robbins), banker M...
2           3 August 2001  ...  In 1947, Andy Dufresne (Tim Robbins), banker M...
3        1 September 2002  ...  In 1947, Andy Dufresne (Tim Robbins), banker M...
4             20 May 2004  ...  In 1947, Andy Dufresne (Tim Robbins), banker M...
...                   ...  ...                                                ...
573901  11 September 1999  ...  The film told chronological format, story told...
573902   4 September 1999  ...  The film told chronological format, story told...
573903     22 August 1999  ...  The film told chronological format, story told...
573904     18 August 1999  ...  The film told chronological format, story told...
573905     29 August 1999  ...  The film told chronological format, story told...

[573906 rows x 12 columns]

In [ ]:
refined_df['review_summary_without_stopwords'] = refined_df['review_summary'].apply(lambda x: ' '.join([word for word in str(x).split() if word not in (stop)]))

In [ ]:
refined_df['plot_summary_without_stopwords'] = refined_df['plot_summary'].apply(lambda x: ' '.join([word for word in str(x).split() if word not in (stop)]))

In [ ]:
refined_df

review_date  ...                     plot_summary_without_stopwords
0        10 February 2006  ...  Chronicles experiences formerly successful ban...
1        6 September 2000  ...  Chronicles experiences formerly successful ban...
2           3 August 2001  ...  Chronicles experiences formerly successful ban...
3        1 September 2002  ...  Chronicles experiences formerly successful ban...
4             20 May 2004  ...  Chronicles experiences formerly successful ban...
...                   ...  ...                                                ...
573901  11 September 1999  ...  Told three perspectives, story bunch young Cal...
573902   4 September 1999  ...  Told three perspectives, story bunch young Cal...
573903     22 August 1999  ...  Told three perspectives, story bunch young Cal...
573904     18 August 1999  ...  Told three perspectives, story bunch young Cal...
573905     29 August 1999  ...  Told three perspectives, story bunch young Cal...

[573906 rows x 14 columns]

In [ ]:
refined_df.to_csv("refined_ds.csv")

In [ ]:
%%bash
zip dataset_rf.zip refined_ds.csv

  adding: refined_ds.csv (deflated 92%)


Day 4

In [ ]:
!unzip 'drive/MyDrive/dataset_rf.zip'

Archive:  drive/MyDrive/dataset_rf.zip
  inflating: refined_ds.csv          


In [ ]:
df = pd.read_csv("refined_ds.csv")

In [ ]:
df.drop(['review_summary','plot_summary'],axis=1,inplace=True)
df

Unnamed: 0  ...                     plot_summary_without_stopwords
0                0  ...  Chronicles experiences formerly successful ban...
1                1  ...  Chronicles experiences formerly successful ban...
2                2  ...  Chronicles experiences formerly successful ban...
3                3  ...  Chronicles experiences formerly successful ban...
4                4  ...  Chronicles experiences formerly successful ban...
...            ...  ...                                                ...
573901      573901  ...  Told three perspectives, story bunch young Cal...
573902      573902  ...  Told three perspectives, story bunch young Cal...
573903      573903  ...  Told three perspectives, story bunch young Cal...
573904      573904  ...  Told three perspectives, story bunch young Cal...
573905      573905  ...  Told three perspectives, story bunch young Cal...

[573906 rows x 13 columns]

In [ ]:
df['genre'] =  df['genre'].apply(lambda x: x.replace('[','').replace(']','').replace("'","").replace(" ",""))
df

Unnamed: 0  ...                     plot_summary_without_stopwords
0                0  ...  Chronicles experiences formerly successful ban...
1                1  ...  Chronicles experiences formerly successful ban...
2                2  ...  Chronicles experiences formerly successful ban...
3                3  ...  Chronicles experiences formerly successful ban...
4                4  ...  Chronicles experiences formerly successful ban...
...            ...  ...                                                ...
573901      573901  ...  Told three perspectives, story bunch young Cal...
573902      573902  ...  Told three perspectives, story bunch young Cal...
573903      573903  ...  Told three perspectives, story bunch young Cal...
573904      573904  ...  Told three perspectives, story bunch young Cal...
573905      573905  ...  Told three perspectives, story bunch young Cal...

[573906 rows x 13 columns]

In [ ]:
cleaned = df.set_index('Unnamed: 0')['genres'].str.split(',', expand=True).stack()
df_dummies = pd.get_dummies(df, columns=['genre'])
df_dummies

KeyError: ignored

In [ ]:
dummies = df['genre'].str.get_dummies(sep=',')
dummies

Action  Adventure  Animation  Biography  ...  Sport  Thriller  War  Western
0            0          0          0          0  ...      0         0    0        0
1            0          0          0          0  ...      0         0    0        0
2            0          0          0          0  ...      0         0    0        0
3            0          0          0          0  ...      0         0    0        0
4            0          0          0          0  ...      0         0    0        0
...        ...        ...        ...        ...  ...    ...       ...  ...      ...
573901       0          0          0          0  ...      0         0    0        0
573902       0          0          0          0  ...      0         0    0        0
573903       0          0          0          0  ...      0         0    0        0
573904       0          0          0          0  ...      0         0    0        0
573905       0          0          0          0  ...      0         0    0        0

[573906 rows x 21 columns]

In [ ]:
final = pd.concat([df,dummies],axis=1)
final

Unnamed: 0        review_date  is_spoiler  ...  Thriller War Western
0                0   10 February 2006        True  ...         0   0       0
1                1   6 September 2000        True  ...         0   0       0
2                2      3 August 2001        True  ...         0   0       0
3                3   1 September 2002        True  ...         0   0       0
4                4        20 May 2004        True  ...         0   0       0
...            ...                ...         ...  ...       ...  ..     ...
573901      573901  11 September 1999       False  ...         0   0       0
573902      573902   4 September 1999       False  ...         0   0       0
573903      573903     22 August 1999       False  ...         0   0       0
573904      573904     18 August 1999       False  ...         0   0       0
573905      573905     29 August 1999       False  ...         0   0       0

[573906 rows x 34 columns]

In [ ]:
final.drop(['genre'],1,inplace=True)
final

Unnamed: 0        review_date  is_spoiler  ...  Thriller War Western
0                0   10 February 2006        True  ...         0   0       0
1                1   6 September 2000        True  ...         0   0       0
2                2      3 August 2001        True  ...         0   0       0
3                3   1 September 2002        True  ...         0   0       0
4                4        20 May 2004        True  ...         0   0       0
...            ...                ...         ...  ...       ...  ..     ...
573901      573901  11 September 1999       False  ...         0   0       0
573902      573902   4 September 1999       False  ...         0   0       0
573903      573903     22 August 1999       False  ...         0   0       0
573904      573904     18 August 1999       False  ...         0   0       0
573905      573905     29 August 1999       False  ...         0   0       0

[573906 rows x 33 columns]

In [ ]:
final['review_date'] =  df['review_date'].apply(lambda x: x[-4:])

In [ ]:
final['release_date'] =  df['release_date'].apply(lambda x: x[:4])

In [ ]:
duration = df['duration']
duration = duration.str.replace("h","")
duration = duration.str.replace("min","")
split_dur = duration.str.split(" ")
split_dur

0         [2, 22]
1         [2, 22]
2         [2, 22]
3         [2, 22]
4         [2, 22]
           ...   
573901    [1, 42]
573902    [1, 42]
573903    [1, 42]
573904    [1, 42]
573905    [1, 42]
Name: duration, Length: 573906, dtype: object

In [ ]:
minutes = []
for i in split_dur:
  if len(i) == 2:
    mins = int(i[0]) * 60 + int(i[1])
  else:
    if int(i[0]) > 6:
      mins = int(i[0])
    else:
      mins = int(i[0]) * 60
  minutes.append(mins)
minutes[-1000]

97

In [ ]:
final['duration'] = minutes
final

Unnamed: 0 review_date  is_spoiler  ...  Thriller War  Western
0                0        2006        True  ...         0   0        0
1                1        2000        True  ...         0   0        0
2                2        2001        True  ...         0   0        0
3                3        2002        True  ...         0   0        0
4                4        2004        True  ...         0   0        0
...            ...         ...         ...  ...       ...  ..      ...
573901      573901        1999       False  ...         0   0        0
573902      573902        1999       False  ...         0   0        0
573903      573903        1999       False  ...         0   0        0
573904      573904        1999       False  ...         0   0        0
573905      573905        1999       False  ...         0   0        0

[573906 rows x 33 columns]

In [ ]:
final.dtypes

Unnamed: 0                            int64
review_date                          object
is_spoiler                             bool
rating_x                              int64
movie_id_y                           object
duration                              int64
rating_y                            float64
release_date                         object
review_without_stopwords             object
description_without_stopwords        object
review_summary_without_stopwords     object
plot_summary_without_stopwords       object
Action                                int64
Adventure                             int64
Animation                             int64
Biography                             int64
Comedy                                int64
Crime                                 int64
Drama                                 int64
Family                                int64
Fantasy                               int64
Film-Noir                             int64
History                         

In [ ]:
final = final.astype({'review_date':'int64','release_date':'int64'})

In [ ]:
final.dtypes

Unnamed: 0                            int64
review_date                           int64
is_spoiler                             bool
rating_x                              int64
movie_id_y                           object
duration                              int64
rating_y                            float64
release_date                          int64
review_without_stopwords             object
description_without_stopwords        object
review_summary_without_stopwords     object
plot_summary_without_stopwords       object
Action                                int64
Adventure                             int64
Animation                             int64
Biography                             int64
Comedy                                int64
Crime                                 int64
Drama                                 int64
Family                                int64
Fantasy                               int64
Film-Noir                             int64
History                         

In [ ]:
final.drop(['Unnamed: 0'],1,inplace=True)
final

review_date  is_spoiler  rating_x  ... Thriller  War  Western
0              2006        True        10  ...        0    0        0
1              2000        True        10  ...        0    0        0
2              2001        True         8  ...        0    0        0
3              2002        True        10  ...        0    0        0
4              2004        True         8  ...        0    0        0
...             ...         ...       ...  ...      ...  ...      ...
573901         1999       False        10  ...        0    0        0
573902         1999       False         9  ...        0    0        0
573903         1999       False         9  ...        0    0        0
573904         1999       False        10  ...        0    0        0
573905         1999       False         8  ...        0    0        0

[573906 rows x 32 columns]

In [ ]:
final.to_csv("final_dataset.csv")

In [ ]:
!zip final_dataset.zip final_dataset.csv

  adding: final_dataset.csv (deflated 92%)


Day 5

In [ ]:
!unzip 'drive/MyDrive/final_dataset.zip'

Archive:  drive/MyDrive/final_dataset.zip
  inflating: final_dataset.csv       


In [ ]:
df = pd.read_csv('final_dataset.csv')

In [ ]:
df["new_column"] = df['review_summary_without_stopwords'].str.replace('[^\w\s]','')
df

Unnamed: 0  ...                                         new_column
0                0  ...           A classic piece unforgettable filmmaking
1                1  ...                   Simply amazing The best film 90s
2                2  ...                      The best story ever told film
3                3  ...                             Busy dying busy living
4                4  ...                  Great story wondrously told acted
...            ...  ...                                                ...
573901      573901  ...  Go gave I wanted 103 minutes action lots super...
573902      573902  ...                                        Right money
573903      573903  ...                 Did somebody say Quentin Tarantino
573904      573904  ...                         innovative rockinI love Go
573905      573905  ...                                          knot bows

[573906 rows x 34 columns]

In [ ]:
df['review_summary_without_stopwords']([y for x in df.values.flatten() for y in str(x).split()]).value_counts()

Lets try fitting models on the non-text variables

In [ ]:
!pip install catboost

     |████████████████████████████████| 67.3MB 64kB/s 


In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import make_moons, make_circles, make_classification
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.gaussian_process import GaussianProcessClassifier
from sklearn.gaussian_process.kernels import RBF
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
from catboost import CatBoostClassifier
from xgboost import XGBClassifier
from sklearn.linear_model import SGDClassifier
from lightgbm import LGBMClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, roc_auc_score
from sklearn.preprocessing import LabelEncoder
from sklearn.inspection import permutation_importance


In [ ]:
classifiers = [[CatBoostClassifier(verbose=0),'CatBoost Classifier'],[XGBClassifier(),'XGB Classifier'], [RandomForestClassifier(),'Random Forest'], 
    [KNeighborsClassifier(), 'K-Nearest Neighbours'], [SGDClassifier(),'SGD Classifier'], [SVC(),'SVC'],[LGBMClassifier(),'LGBM Classifier'],
              [GaussianNB(),'GaussianNB'],[DecisionTreeClassifier(),'Decision Tree Classifier'],[LogisticRegression(),'Logistic Regression']]

In [ ]:
names = ["Nearest Neighbors", "Linear SVM", "RBF SVM", "Gaussian Process",
         "Decision Tree", "Random Forest", "Neural Net", "AdaBoost",
         "Naive Bayes", "QDA"]

In [ ]:
X = df.copy()
y = X['is_spoiler']
X = X.drop(["review_without_stopwords","description_without_stopwords","review_summary_without_stopwords","plot_summary_without_stopwords"],1)

In [ ]:
X.drop(["Unnamed: 0"], 1,inplace=True)

In [ ]:
X = X.drop(['is_spoiler'],1)
X

review_date  rating_x movie_id_y  ...  Thriller  War  Western
0              2006        10  tt0111161  ...         0    0        0
1              2000        10  tt0111161  ...         0    0        0
2              2001         8  tt0111161  ...         0    0        0
3              2002        10  tt0111161  ...         0    0        0
4              2004         8  tt0111161  ...         0    0        0
...             ...       ...        ...  ...       ...  ...      ...
573901         1999        10  tt0139239  ...         0    0        0
573902         1999         9  tt0139239  ...         0    0        0
573903         1999         9  tt0139239  ...         0    0        0
573904         1999        10  tt0139239  ...         0    0        0
573905         1999         8  tt0139239  ...         0    0        0

[573906 rows x 27 columns]

In [ ]:
X.drop(['movie_id_y'],1,inplace=True)

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 42)


In [ ]:
def predict(X_train, y_train,X_test,y_test):
    for cls in classifiers:
        model = cls[0]
        model.fit(X_test, y_test)

        y_pred = model.predict(X_test)
        print(cls[1])
        print("Accuracy : ", accuracy_score(y_test, y_pred) *  100)
        print("Roc AUC score: ", roc_auc_score(y_test,y_pred))

In [ ]:
X_train.dtypes

review_date       int64
rating_x          int64
duration          int64
rating_y        float64
release_date      int64
Action            int64
Adventure         int64
Animation         int64
Biography         int64
Comedy            int64
Crime             int64
Drama             int64
Family            int64
Fantasy           int64
Film-Noir         int64
History           int64
Horror            int64
Music             int64
Musical           int64
Mystery           int64
Romance           int64
Sci-Fi            int64
Sport             int64
Thriller          int64
War               int64
Western           int64
dtype: object

In [ ]:
predict(X_train, y_train,X_test,y_test)


CatBoost Classifier


TypeError: ignored

In [ ]:
lr = LogisticRegression()
lr.fit(X_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [ ]:
y_pred = lr.predict(X_test)
print("Logistic Regression")
print("Accuracy : ", accuracy_score(y_test, y_pred) *  100)
print("Roc AUC score: ", roc_auc_score(y_test,y_pred))

Logistic Regression
Accuracy :  73.79161816876572
Roc AUC score:  0.5000132964578237


In [ ]:
rf = RandomForestClassifier()
rf.fit(X_train,y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [ ]:
y_pred = rf.predict(X_test)
print("Random Forest")
print("Accuracy : ", accuracy_score(y_test, y_pred) *  100)
print("Roc AUC score: ", roc_auc_score(y_test,y_pred))

Random Forest
Accuracy :  70.88453201558438
Roc AUC score:  0.5372453219755127


In [ ]:
xgb = XGBClassifier()
xgb.fit(X_train,y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=1, missing=None, n_estimators=100, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [ ]:
y_pred = xgb.predict(X_test)
print("XGBoost")
print("Accuracy : ", accuracy_score(y_test, y_pred) *  100)
print("Roc AUC score: ", roc_auc_score(y_test,y_pred))

XGBoost
Accuracy :  73.79092119294381
Roc AUC score:  0.5


In [ ]:
gnb = GaussianNB()
gnb.fit(X_train,y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

In [ ]:
y_pred = gnb.predict(X_test)
print("Gaussian Naive Bayes")
print("Accuracy : ", accuracy_score(y_test, y_pred) *  100)
print("Roc AUC score: ", roc_auc_score(y_test,y_pred))

Gaussian Naive Bayes
Accuracy :  64.86056998682716
Roc AUC score:  0.5738662778432911


In [ ]:
lgbm = LGBMClassifier()
lgbm.fit(X_train,y_train)

LGBMClassifier(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
               importance_type='split', learning_rate=0.1, max_depth=-1,
               min_child_samples=20, min_child_weight=0.001, min_split_gain=0.0,
               n_estimators=100, n_jobs=-1, num_leaves=31, objective=None,
               random_state=None, reg_alpha=0.0, reg_lambda=0.0, silent=True,
               subsample=1.0, subsample_for_bin=200000, subsample_freq=0)

In [ ]:
y_pred = lgbm.predict(X_test)
print("Light GBM")
print("Accuracy : ", accuracy_score(y_test, y_pred) *  100)
print("Roc AUC score: ", roc_auc_score(y_test,y_pred))

Light GBM
Accuracy :  73.91080103431213
Roc AUC score:  0.5054164344744163


In [ ]:
knn = KNeighborsClassifier()
knn.fit(X_train,y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

In [ ]:
y_pred = knn.predict(X_test)
print("K Nearest Neighbors")
print("Accuracy : ", accuracy_score(y_test, y_pred) *  100)
print("Roc AUC score: ", roc_auc_score(y_test,y_pred))

K Nearest Neighbors
Accuracy :  69.16997149368889
Roc AUC score:  0.5421408033849902


In [ ]:
mlp = MLPClassifier()
mlp.fit(X_train,y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:573: UserWarning: Training interrupted by user.
  warnings.warn("Training interrupted by user.")


MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(100,), learning_rate='constant',
              learning_rate_init=0.001, max_fun=15000, max_iter=200,
              momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
              power_t=0.5, random_state=None, shuffle=True, solver='adam',
              tol=0.0001, validation_fraction=0.1, verbose=False,
              warm_start=False)

In [ ]:
y_pred = mlp.predict(X_test)
print("MLP Classifier")
print("Accuracy : ", accuracy_score(y_test, y_pred) *  100)
print("Roc AUC score: ", roc_auc_score(y_test,y_pred))

MLP Classifier
Accuracy :  73.79092119294381
Roc AUC score:  0.5


In [ ]:
cat = CatBoostClassifier()
le = LabelEncoder()
y_enc_train = le.fit_transform(y_train)
cat.fit(X_train,y_enc_train)

Learning rate set to 0.137281
0:	learn: 0.6481301	total: 143ms	remaining: 2m 23s
1:	learn: 0.6159434	total: 291ms	remaining: 2m 25s
2:	learn: 0.5937506	total: 426ms	remaining: 2m 21s
3:	learn: 0.5778389	total: 581ms	remaining: 2m 24s
4:	learn: 0.5665350	total: 722ms	remaining: 2m 23s
5:	learn: 0.5593394	total: 870ms	remaining: 2m 24s
6:	learn: 0.5532265	total: 1.02s	remaining: 2m 25s
7:	learn: 0.5490207	total: 1.17s	remaining: 2m 25s
8:	learn: 0.5459904	total: 1.33s	remaining: 2m 26s
9:	learn: 0.5433015	total: 1.49s	remaining: 2m 27s
10:	learn: 0.5413106	total: 1.63s	remaining: 2m 26s
11:	learn: 0.5400327	total: 1.78s	remaining: 2m 26s
12:	learn: 0.5388702	total: 1.93s	remaining: 2m 26s
13:	learn: 0.5382751	total: 2.07s	remaining: 2m 25s
14:	learn: 0.5375771	total: 2.21s	remaining: 2m 25s
15:	learn: 0.5371854	total: 2.34s	remaining: 2m 24s
16:	learn: 0.5367109	total: 2.49s	remaining: 2m 24s
17:	learn: 0.5361952	total: 2.64s	remaining: 2m 24s
18:	learn: 0.5360151	total: 2.76s	remaining:

In [ ]:
y_enc_test = le.transform(y_test)
y_pred = cat.predict(X_test)
print("CatBoost Classifier")
print("Accuracy : ", accuracy_score(y_enc_test, y_pred) *  100)
print("Roc AUC score: ", roc_auc_score(y_enc_test,y_pred))

CatBoost Classifier
Accuracy :  73.93937704301038
Roc AUC score:  0.5162844666408463


Gaussian NB performs the best so far. Lets see its class weights

In [ ]:
imps = permutation_importance(gnb, X_test, y_test)
print(imps.importances_mean)


[ 0.02028339  0.00277536  0.00022303 -0.00042237 -0.00247287 -0.00185674
  0.00038891  0.00067467 -0.00314057 -0.00124201  0.00252166 -0.00082661
 -0.00013661 -0.00110959 -0.00011012 -0.00294681 -0.00091025 -0.00121134
 -0.0001394  -0.00197105 -0.00316706  0.00021049 -0.00125874 -0.00163789
 -0.00135492  0.0002537 ]


In [ ]:
imps_train = permutation_importance(gnb, X_train, y_train)
print(imps.importances_mean)


[ 0.02028339  0.00277536  0.00022303 -0.00042237 -0.00247287 -0.00185674
  0.00038891  0.00067467 -0.00314057 -0.00124201  0.00252166 -0.00082661
 -0.00013661 -0.00110959 -0.00011012 -0.00294681 -0.00091025 -0.00121134
 -0.0001394  -0.00197105 -0.00316706  0.00021049 -0.00125874 -0.00163789
 -0.00135492  0.0002537 ]


In [ ]:
importance = [x for x in (imps.importances_mean,X_train.columns)]
importance

[array([ 0.02028339,  0.00277536,  0.00022303, -0.00042237, -0.00247287,
        -0.00185674,  0.00038891,  0.00067467, -0.00314057, -0.00124201,
         0.00252166, -0.00082661, -0.00013661, -0.00110959, -0.00011012,
        -0.00294681, -0.00091025, -0.00121134, -0.0001394 , -0.00197105,
        -0.00316706,  0.00021049, -0.00125874, -0.00163789, -0.00135492,
         0.0002537 ]),
 Index(['review_date', 'rating_x', 'duration', 'rating_y', 'release_date',
        'Action', 'Adventure', 'Animation', 'Biography', 'Comedy', 'Crime',
        'Drama', 'Family', 'Fantasy', 'Film-Noir', 'History', 'Horror', 'Music',
        'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Sport', 'Thriller', 'War',
        'Western'],
       dtype='object')]

In [ ]:
print(sorted(imps.importances_mean))

[-0.0031670581347533087, -0.003140573053520801, -0.002946813775030166, -0.0024728702161322057, -0.0019710476243579135, -0.0018567435895649354, -0.0016378931814855723, -0.0013549209977906207, -0.001258738334367182, -0.0012420109146414004, -0.0012113439784773971, -0.0011095855084787276, -0.0009102504234128394, -0.0008266133247837981, -0.0004223673480767243, -0.00013939516438175037, -0.000136607261094146, -0.00011012217986161588, 0.00021048669821643884, 0.0002230322630107695, 0.0002536991991747728, 0.00038891250862507223, 0.0006746725956076283, 0.002521658523665793, 0.0027753577228405656, 0.020283390369188047]


Highest value features are review_date, crime, IMDB rating, Animation

Lowest value features are Romance, Biography, History

In [ ]:
spoilers = df[df["is_spoiler"]]
spoilers

Unnamed: 0  review_date  is_spoiler  ...  Thriller War  Western
0                0         2006        True  ...         0   0        0
1                1         2000        True  ...         0   0        0
2                2         2001        True  ...         0   0        0
3                3         2002        True  ...         0   0        0
4                4         2004        True  ...         0   0        0
...            ...          ...         ...  ...       ...  ..      ...
573640      573640         1999        True  ...         0   0        0
573641      573641         2006        True  ...         0   0        0
573642      573642         2000        True  ...         0   0        0
573643      573643         2014        True  ...         0   0        0
573644      573644         2003        True  ...         0   0        0

[150924 rows x 33 columns]

In [ ]:
spoilers_cat = spoilers.drop(["Unnamed: 0", 'review_date','is_spoiler','rating_x','movie_id_y','duration','rating_y','release_date','review_without_stopwords','description_without_stopwords','review_summary_without_stopwords','plot_summary_without_stopwords'],1)


In [ ]:
spoilers_cat.value_counts().head(10)

Action  Adventure  Animation  Biography  Comedy  Crime  Drama  Family  Fantasy  Film-Noir  History  Horror  Music  Musical  Mystery  Romance  Sci-Fi  Sport  Thriller  War  Western
1       1          0          0          0       0      0      0       0        0          0        0       0      0        0        0        1       0      0         0    0          12959
                                                                       1        0          0        0       0      0        0        0        0       0      0         0    0           6682
0       0          0          0          0       1      1      0       0        0          0        0       0      0        0        0        0       0      0         0    0           4450
                                                 0      1      0       0        0          0        0       0      0        0        1        0       0      0         0    0           4211
                                                                

In [ ]:
spoilers_cat['Action'].value_counts()

0    97026
1    53898
Name: Action, dtype: int64

In [ ]:
spoilers_cat['Adventure'].value_counts()

0    97189
1    53735
Name: Adventure, dtype: int64

In [ ]:
spoilers_cat['Animation'].value_counts()

0    144090
1      6834
Name: Animation, dtype: int64

In [ ]:
spoilers_cat['Biography'].value_counts()

0    142932
1      7992
Name: Biography, dtype: int64

In [ ]:
spoilers_cat['Comedy'].value_counts()

0    121612
1     29312
Name: Comedy, dtype: int64

In [ ]:
spoilers_cat['Crime'].value_counts()

0    123911
1     27013
Name: Crime, dtype: int64

In [ ]:
spoilers_cat['Drama'].value_counts()

0    76693
1    74231
Name: Drama, dtype: int64

In [ ]:
spoilers_cat['Family'].value_counts()

0    140222
1     10702
Name: Family, dtype: int64

In [ ]:
spoilers_cat['Fantasy'].value_counts()

0    129230
1     21694
Name: Fantasy, dtype: int64

In [ ]:
spoilers_cat['Film-Noir'].value_counts()

0    150436
1       488
Name: Film-Noir, dtype: int64

In [ ]:
History  Horror  Music  Musical  Mystery  Romance  Sci-Fi  Sport  Thriller  War  Western

In [ ]:
spoilers_cat['History'].value_counts()

0    145256
1      5668
Name: History, dtype: int64

In [ ]:
spoilers_cat['Horror'].value_counts()

0    137045
1     13879
Name: Horror, dtype: int64

In [ ]:
spoilers_cat['Music'].value_counts()

0    149086
1      1838
Name: Music, dtype: int64

In [ ]:
spoilers_cat['Musical'].value_counts()

0    149804
1      1120
Name: Musical, dtype: int64

In [ ]:
spoilers_cat['Mystery'].value_counts()

0    132381
1     18543
Name: Mystery, dtype: int64

In [ ]:
for i in spoilers_cat.columns:
  print(spoilers_cat[i].value_counts())
  print(spoilers_cat[i].value_counts()[1]/(spoilers_cat[i].value_counts()[1]+spoilers_cat[i].value_counts()[0]) * 100)

0    97026
1    53898
Name: Action, dtype: int64
35.7120139937982
0    97189
1    53735
Name: Adventure, dtype: int64
35.60401261562111
0    144090
1      6834
Name: Animation, dtype: int64
4.528106861731732
0    142932
1      7992
Name: Biography, dtype: int64
5.295380456388646
0    121612
1     29312
Name: Comedy, dtype: int64
19.421695687895895
0    123911
1     27013
Name: Crime, dtype: int64
17.898412445999313
0    76693
1    74231
Name: Drama, dtype: int64
49.18435768996316
0    140222
1     10702
Name: Family, dtype: int64
7.090986191725637
0    129230
1     21694
Name: Fantasy, dtype: int64
14.374122074686596
0    150436
1       488
Name: Film-Noir, dtype: int64
0.3233415493890965
0    145256
1      5668
Name: History, dtype: int64
3.7555325859372926
0    137045
1     13879
Name: Horror, dtype: int64
9.196019188465717
0    149086
1      1838
Name: Music, dtype: int64
1.217831491346638
0    149804
1      1120
Name: Musical, dtype: int64
0.7420953592536641
0    132381
1     18543

Day 6

In [ ]:
!unzip 'drive/MyDrive/final_dataset.zip'

Archive:  drive/MyDrive/final_dataset.zip
  inflating: final_dataset.csv       


In [ ]:
df = pd.read_csv('final_dataset.csv')

In [ ]:
import re


def clean_text(text):
    """
    Applies some pre-processing on the given text.

    Steps :
    - Removing HTML tags
    - Removing punctuation
    - Lowering text
    """
    
    # remove HTML tags
    text = re.sub(r'<.*?>', '', text)
    
    # remove the characters [\], ['] and ["]
    text = re.sub(r"\\", "", text)    
    text = re.sub(r"\'", "", text)    
    text = re.sub(r"\"", "", text)    
    
    # convert text to lowercase
    text = text.strip().lower()
    
    # replace punctuation characters with spaces
    filters='!"\'#$%&()*+,-./:;<=>?@[\\]^_`{|}~\t\n'
    translate_dict = dict((c, " ") for c in filters)
    translate_map = str.maketrans(translate_dict)
    text = text.translate(translate_map)

    return text


In [ ]:
df["review_summary_without_stopwords"] = df["review_summary_without_stopwords"].astype(str)

In [ ]:
from sklearn.metrics import accuracy_score
from sklearn.svm import LinearSVC
from sklearn.feature_extraction.text import CountVectorizer



# Transform each text into a vector of word counts
vectorizer = CountVectorizer(stop_words="english",
                             preprocessor=clean_text)

summary_bow = vectorizer.fit_transform(df["review_summary_without_stopwords"])    

In [ ]:
summary_bow.toarray()